Console/UI/Pages/EventsLog.xaml.cs

```diff
@@ -126,94 +140,114 @@ public EventsLog()
+        private void StopHandlingSecurityLogEvents()
+        {
+            PauseHandlingSecurityLogEvents();
+            securityLog?.Dispose();
         }
 
         private void SecurityLog_EntryWritten(object sender, EntryWrittenEventArgs e)
         {
+            //LogHelper.Debug($"EntryWritten event ...");
             if (!FirewallHelper.IsEventAccepted(e.Entry))
             {
                 return;
             }
 
-            LogEntryViewModel entry = EntryViewFromLogEntry(e.Entry);
-            if (entry != null)
+            lock (EntriesLock)
             {
-                Dispatcher.Invoke(() =>
+                if (e.Entry != null && e.Entry.Index > SecurityLogEntryIndex_last)
                 {
-                    if (!TargetCollectionNewEntries.Contains(entry, EntryComparer.Instance))
+                    LogEntryViewModel entry = EntryViewFromLogEntry(e.Entry);
+                    if (entry != null)
                     {
-                        TargetCollectionNewEntries.Add(entry);
+                        Dispatcher.Invoke(() =>
+                        {
+                            Dispatcher.Invoke(() => { LogEntries.Add(entry); ScanProgress = LogEntries.Count; });
+                        });
                     }
-                });
+                }
+                if (ScanProgressMax < LogEntries.Count)
+                {
+                    // overflow handling
+                    LogEntries.RemoveAt(0);
+                }
+                SecurityLogEntryIndex_last = securityLog.Entries.Count - 1;
             }
         }
 
-        private static bool TCPFilter(object le) => ((LogEntryViewModel)le).Protocol == "TCP";
-
-        private void InitEventLog(Action<int> progressCallback)
+        /*
+         * Initializes the LogEntries list to show the most recent events
+         */
+        private void InitEventLog(int lastIndex, Action<int> progressCallback)
         {
+            int i = 0;
             try
             {
-                Dispatcher.Invoke(() => LogEntries.Clear());
-
-                ScanProgressMax = securityLog.Entries.Count;
-
-                // Have to use a Enumerable.Range to enumerate the entries starting from the last one, since calling Reverse on the initial collection will first enumerate all entries
-                var entries = Enumerable.Range(0, ScanProgressMax)
-                                        .Select(i => { progressCallback?.Invoke(i + 1); return securityLog.Entries[ScanProgressMax - i - 1]; })
-                                        .Where(FirewallHelper.IsEventAccepted)
-                                        .Select(EntryViewFromLogEntry)
-                                        .Where(entry => entry != null)
-                                        .Select(entry => { Dispatcher.Invoke(() => { LogEntries.Add(entry); }); return entry; })
-                                        .ToList();
+                LogHelper.Debug($"InitEventLog ...");

```

Comment:
I liked my Linq styled code snippet, but, well... ;)

Reply:
Can image you've been proud of 😄 ... took me a while to figure out what it does (not being so much used to linq) and debugging it is a pain.

Reply:
It does (almost) the same as a standard loop, except it's more "fluent". Regarding debugging, you can set breakpoints in (or step into) the lambda methods or delegate, it doesn't change anything in fact :)

Reply:
Yes almost 😉, but the fluent nature also has some drawbacks in this case, from my point of view:

- it's less flexible than a for loop e.g. placing the lock 

- don't like, that a list has to be produced which is then not used e.g:
``` var entries = ...toList()``` - also it toList() is ommited, it compiles but behaves differently - if I'm correct though 🤔 

- debugging - well maybe I'm not used to do it properly 🤔

Reply:
Hi @harrwiss,

You can also place a lock in a lambda :) _collection.Select(item => { lock(collectionLocker) { collectionLocker.Add(item); } })_ would do it.
And if it looks too messy, _collection.Select(AddItemWithLock)_ should be more readable (with AddItemWithLock(TItem item) being either a local method or a class-level one).

For the ToList part, in fact a LINQ Enumerable isn't evaluated (enumerated) until you call ToList or ToArray (or of course enumerate it directly using a for* loop or through its enumerator).
Writing _collection.Select(item => item.Property)_ only prepares the query (parsing it into an expression tree) but doesn't actually do anything (same goes for Streams in Java where _Collect(Collectors.toList())_ is the ugly equivalent for ToList(), for instance).
So _collection.Select(...).Where(...).Union(...).Concat(...).WhatEver(...)_ doesn't enumerate the collection for each extension method called either, it just builds a query (can envision it as a hibernate query in Java, where the SQL statement is built but not executed yet).

**Regarding the fact that it's not really smart to create a List while we don't need it, you're totally right.** Wrote this quickly, while I have a custom extension method to actually enumerate while not creating anything out of it (to keep it consistent with what we actually want to do), like a standard for/foreach.
**But in fact the best way would have been to stop the LINQ syntax after the last "where", and keep a foreach on the resulting collection (for the Dispatcher invocation). Would have been clearer that way that we are not actually projecting the collection items, but only using them direcly.**

Note: another way to write LINQ is like _from _collection_ select Property etc..._ (which makes that "defered" part clearer I guess) but I really dislike it. Kinda mimicking HQL or SQL while being way more complex when it comes to groupbys or joins...

Debugging: since enumeration occurs only when needed, breakpoints won't be hit before the first ToList/ToArray/explicit enumeration. Could explain why it didn't behave as you expected?

Reply:
> But in fact the best way would have been to stop the LINQ syntax after the last "where", and keep a foreach on the resulting collection (for the Dispatcher invocation). Would have been clearer that way that we are not actually projecting the collection items, but only using them direcly.

👍 thx for all the explanations... and I'm not against LINQ - in fact I like it, when used for the right purpose 😄  For instance, its very good for filtering and transforming a data stream before processing but things should not be mixed up (as you described in the quote above)